core/cp4d_authenticator.go

```diff
@@ -144,24 +151,80 @@ func (authenticator *CloudPakForDataAuthenticator) Authenticate(request *http.Re
 }
 
 // getToken: returns an access token to be used in an Authorization header.
-// Whenever a new token is needed (when a token doesn't yet exist, or the existing token has expired),
-// a new access token is fetched from the token server.
+// Whenever a new token is needed (when a token doesn't yet exist, needs to be refreshed,
+// or the existing token has expired), a new access token is fetched from the token server.
 func (authenticator *CloudPakForDataAuthenticator) getToken() (string, error) {
 	if authenticator.tokenData == nil || !authenticator.tokenData.isTokenValid() {
-		tokenResponse, err := authenticator.requestToken()
+		// synchronously request the token
+		err := authenticator.synchronizedRequestToken()
 		if err != nil {
 			return "", err
 		}
-
-		authenticator.tokenData, err = newCp4dTokenData(tokenResponse)
-		if err != nil {
-			return "", err
+	} else if authenticator.tokenData != nil && authenticator.needsRefresh() {
+		// If refresh needed, kick off a go routine in the background to get a new token
+		ch := make(chan error)
+		go func() {
+			ch <- authenticator.getTokenData()
+		}()
+		select {
+		case err := <-ch:
+			if err != nil {
+				return "", err
+			}
+		default:
 		}
 	}
 
+	// return an error if the access token is not valid or was not fetched
+	if authenticator.tokenData == nil || authenticator.tokenData.AccessToken == "" {
+		return "", fmt.Errorf("Error while trying to get access token")
+	}
+
 	return authenticator.tokenData.AccessToken, nil
 }
 
+// synchronizedRequestToken: synchronously checks if the current token in cache
+// is valid. If token is not valid or does not exist, it will fetch a new token
+// and set the tokenRefreshTime
+func (authenticator *CloudPakForDataAuthenticator) synchronizedRequestToken() error {
+	cp4dMuOne.Lock()
+	defer cp4dMuOne.Unlock()
+	// if cached token is still valid, then just continue to use it
+	if authenticator.tokenData != nil && authenticator.tokenData.isTokenValid() {
+		return nil
+	}
+
+	err := authenticator.getTokenData()
```

Comment:
I think it is very dangerous to perform a network call while holding a mutex.  We should find some other solution.

Reply:
Thinking about this further ... I suppose the other languages must do something similar.  Hmmm.

Reply:
In the Java core, we do this inside the TokenRequestBasedAuthenticator's getToken() method:
- if we have no token at all, or the current token has expired, then we invoke a blocking request (synchronized) to the token server
- otherwise we have a valid token and we use that as-is, but if the token is within the refresh window (last 20% of lifetime I think), then we ALSO start a refresh in a separate thread.

Not sure what else we can do to improve on that

Reply:
@mkistler - I think the only other thing I can think of to use in place of the mutex is a blocking go routine.. Though I remember reading that under hood it still uses a mutex. I'll try to find that text again

Reply:
My concern here is ... what happens if the _one_ token request we make "gets lost" -- not unheard of for a network request.  Does the client-side time out after a while and simply fail the request ... allowing another to try the request again?  If that's the case then this is probably fine.

Reply:
whatever happens in the Java core would also happen here, so yeah I guess a timeout would occur.   If there is some other more fault-tolerant pattern we should be using in the Node, Java, Go cores (at a minimum) to avoid this, then we'll need to put our thinking caps on and come up with a better design.

Reply:
Agreed.  I was being overly cautious.  The timeout will cover us so this should be fine.